src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -246,19 +246,24 @@ MemoryScope acquire() {
         }
 
         @Override
-        MemoryScope dup(Thread owner) {
-            Objects.requireNonNull(owner, "owner");
-            return closeOrDup(owner);
+        MemoryScope dup(Thread newOwner) {
+            Objects.requireNonNull(newOwner, "newOwner");
+            // pre-allocate duped scope so we don't get OOME later and be left with this scope closed
+            var duped = new Root(newOwner, ref, cleanupAction);
+            justClose();
+            return duped;
         }
 
         @Override
         void close() {
-            closeOrDup(null);
+            justClose();
```

Comment:
This is exactly what I had in mind - I think the code is cleaner. Perhaps rename `justClose` to `doClose` (and make `doClose` private) - but that's minor - I'll leave that up to you.

Reply:
I think justClose() tells more that doClose() and private it is.